sdk/storage/Azure.Storage.DataMovement/src/DataTransferStatus.cs

```diff
@@ -1,66 +1,147 @@
         /// </summary>
-        Completed = 4,
+        protected internal DataTransferStatus()
+        {
+            State = DataTransferState.Queued;
+            HasFailedItems = false;
+            HasSkippedItems = false;
+        }
 
         /// <summary>
-        /// The Job has been completed with at least one skipped transfer.
+        /// Constructor to have a custom state, failure state, and skipped state.
         /// </summary>
-        CompletedWithSkippedTransfers = 5,
+        protected internal DataTransferStatus(DataTransferState state, bool hasFailureItems, bool hasSkippedItems)
+        {
+            State = state;
+            HasFailedItems = hasFailureItems;
+            HasSkippedItems = hasSkippedItems;
+        }
+
+        internal bool IsCompletedWithFailedItems => State.Equals(DataTransferState.Completed) && HasFailedItems;
+        internal bool IsCompletedWithSkippedItems => State.Equals(DataTransferState.Completed) && HasSkippedItems;
 
         /// <summary>
-        /// The Job has been completed with at least one failed transfer.
+        /// Accordingly update the <see cref="HasFailedItems"/> to true. If already set to true, nothing will happen.
+        ///
+        /// This should only be triggered when a failed item has been seen.
+        /// </summary>
+        /// <returns>True if <see cref="HasFailedItems"/> was updated. False otherwise.</returns>
+        internal bool TrySetFailedItem()
+        {
+            if (!HasFailedItems)
+            {
+                HasFailedItems = true;
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Accordingly update the <see cref="HasSkippedItems"/> to true. If already set to true, nothing will happen.
+        ///
+        /// This should only be triggered when a skipped item has been seen.
+        /// </summary>
+        /// /// <returns>True if <see cref="HasSkippedItems"/> was updated. False otherwise.</returns>
+        internal bool TrySetSkippedItem()
+        {
+            if (!HasSkippedItems)
+            {
+                HasSkippedItems = true;
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Accordingly update the <see cref="State"/>. If the current State is the same as the parameter,
+        /// then nothing will happen.
+        ///
+        /// This should only be triggered when the state updates.
         /// </summary>
-        CompletedWithFailedTransfers = 6,
+        /// <returns>True if <see cref="State"/> was updated. False otherwise.</returns>
+        internal bool TrySetTransferStateChange(DataTransferState state)
+        {
+            lock (_stateLock)
+            {
+                if (state != DataTransferState.None &&
+                    State != state)
+                {
+                    State = state;
+                    return true;
+                }
+            }
```

Comment:
Since enums are just ints, you can probably use Interlocked.CompareExchange() here and eliminate the final lock.